TweetyBird/twitter_utils.py

```diff
@@ -0,0 +1,152 @@
+import re
+
+from replit import db
+import tweepy
+
+from .discord_utils import send_discord_message
+from .settings import Twitter_API_PK, Twitter_API_SK, Twitter_Access_Token, Twitter_Access_Secret
+
+
+# Modify the default Twitter Stream, and overwrite the default on_status call
+class TwitterStreamListener(tweepy.StreamListener):
+    def on_status(self, status):
+        send_discord_message(format_tweet(status))
+
+
+# Authenticate with Twitter
+auth = tweepy.OAuthHandler(Twitter_API_PK, Twitter_API_SK)
+auth.set_access_token(Twitter_Access_Token, Twitter_Access_Secret)
+api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)
+try:
+    api.verify_credentials()
+    print("Authentication OK")
+except Exception:
+    print("Error during authentication")
+
+# Tweepy Stream setup
+myStreamListener = TwitterStreamListener()
+myStream = tweepy.Stream(auth=api.auth, listener=myStreamListener)
+
+# Filter requires list of ID's and a filter level (none, low, medium, high)
+myStream.filter(follow=db["TwitterFollows"], is_async=True, filter_level="medium")
+
+#Regex Cached URL Finder
+p = re.compile('(https://t.co/[a-zA-Z0-9]{10})')
+
+
+def update_following(twitter_user):
+    """
+    If the provided user is a valid Twitter user:
+      - Add the user to the following
+      - Init db if this is the first user
+    """
+    msg = "Not a valid Twitter Account"
+    if is_valid_twitter_user(twitter_user):
+        twitter_user_id_string = api.get_user(twitter_user).id_str
+        # Ensure database key exists
+        if "TwitterFollows" in db.keys():
+            following = db["TwitterFollows"]
+            # Check for duplicates
+            if twitter_user_id_string in following:
+                msg = "Twitter User already in Subscription List"
+            else:
+                following.append(twitter_user_id_string)
+                db["TwitterFollows"] = following
+                msg = 'Adding ' + twitter_user + ' to Follow List'
```

Comment:
I would change the db structure to be a dictionary of dictionaries instead of a dictionary of lists. That would simplify these kinds of operations a ton and make them more performant.

For example, current implementation:

```
{
  "TwitterFollows": ['uid1', 'uid2', 'uid3']
}
```

means you (python) have to check the whole list every time

vs

```
{
  "TwitterFollows": {
     "uid1": True
     "uid2": True
  }
}
```

means you (python) can hash the key and get the result in one operation. Big O nerd stuff -- https://bradfieldcs.com/algos/analysis/performance-of-python-types/

The above code section could turn into

```
following = db.get("TwitterFollows", {})
if twitter_user_id_string in following:
    msg = "Twitter user already in subscription list"
else:
    following[twitter_user_id_string] = True  # or any other value, we don't really need the value
    db["TwitterFollows] = following
```